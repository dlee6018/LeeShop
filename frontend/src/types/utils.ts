import { RouteComponentProps } from "react-router-dom";
interface Props extends RouteComponentProps {

}
interface OrderItems{
    name: string,
    qty: number,
    image: string,
    price: number,
    products: string
}

export interface IShippingAddress{
    address: string,
    city: string,
    postalCode: string,
    country: string
}
interface IPaymentResult{
    id: string,
    status: string,
    update_time: string,
    email_address: string
}
interface Review {
    _id: string,
    name: string,
    rating: number,
    comment: string,
    user: string,
    createdAt: string,
}

export interface IOrder{
    user: IUserInfo,
    orderItems: Array<OrderItems>,
    shippingAddress: IShippingAddress,
    paymentMethod: string,
    paymentResult: IPaymentResult,
    taxPrice: number,
    shippingPrice: number,
    totalPrice: number,
    isPaid: boolean,
    paidAt: Date,
    isDelivered: boolean,
    deliveredAt: Date
}


export interface IProduct{
    _id: string,
    name: string,
    image: string,
    brand: string,
    category: string,
    description: string,
    reviews?: Array<Review> ,
    rating?: number,
    numReviews?: number,
    price: number,
    countInStock: number,
}

export interface IProductDetails{

}

export interface IUserInfo {
    name: string,
    email:string,
    password: string,
    isAdmin: boolean,
    _id: string,
    token: string,
}

