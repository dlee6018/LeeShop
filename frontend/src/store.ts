import { combineReducers } from "redux";
import { useDispatch } from "react-redux";
import { configureStore} from "@reduxjs/toolkit";
import {
  productListReducer,
  productDetailsReducer,
  productDeleteReducer,
  productCreateReducer,
  productUpdateReducer,
  productReviewCreateReducer,
  productTopRatedReducer,
} from "./reducers/productReducers";
import { cartReducer } from "./reducers/cartReducers";
import {
  userLoginReducer,
  userRegisterReducer,
  userDetailsReducer,
  userUpdateProfileReducer,
  userListReducer,
  userDeleteReducer,
  userUpdateReducer,
} from "./reducers/userReducers";
import {
  orderCreateReducer,
  orderDetailsReducer,
  orderPayReducer,
  orderDeliverReducer,
  orderListMyReducer,
  orderListReducer,
} from "./reducers/orderReducers";
import productsReducer from "./features/products/productSlice";
import usersReducer from "./features/users/userSlice";
import { IUserInfo, IShippingAddress } from "./types/utils";

const rootReducer = combineReducers({
  products: productsReducer,
  users: usersReducer,
  productReviewCreate: productReviewCreateReducer,
  productTopRated: productTopRatedReducer,
  cart: cartReducer,
  userRegister: userRegisterReducer,
  userUpdateProfile: userUpdateProfileReducer,
  userList: userListReducer,
  userDelete: userDeleteReducer,
  userUpdate: userUpdateReducer,
  orderCreate: orderCreateReducer,
  orderDetails: orderDetailsReducer,
  orderPay: orderPayReducer,
  orderDeliver: orderDeliverReducer,
  orderListMy: orderListMyReducer,
  orderList: orderListReducer,

});


// export interface IUser{
//   name: string,
//   email:string,
//   password: string,
//   isAdmin: boolean
// }
// interface shippingAddress{
//   address: string,
//   city: string,
//   postalCode: string,
//   country: string
// }

const cartItemsFromStorage = JSON.parse(
  localStorage.getItem("cartItems") || "[]"
);

const userInfoFromStorage = JSON.parse(
  localStorage.getItem("userInfo") || "[]"
);


const shippingAddressFromStorage = JSON.parse(
  localStorage.getItem("shippingAddress") || "[]"
);

const preloadedState = {
  // cart: {
  //   cartItems: cartItemsFromStorage,
  //   shippingAddress: shippingAddressFromStorage,
  // },
  users: { 
    status: null,
    error:null,
    userInfo: userInfoFromStorage },
};


const store = configureStore({
  reducer: rootReducer,
  preloadedState,
});

export default store;
export type RootState = ReturnType<typeof store.getState>
export type AppDispatch = typeof store.dispatch
export const useAppDispatch: () => AppDispatch = useDispatch
