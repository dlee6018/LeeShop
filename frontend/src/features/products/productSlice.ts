import { IProduct } from './../../types/utils';
import { RootState } from './../../store';
import {
  createSlice,
  createAsyncThunk,
} from "@reduxjs/toolkit";
import axios from "axios";

export const getAllProducts = createAsyncThunk<any, {keyword: string, pageNumber: number}, {}>(
  "product/allProducts",
  async ({keyword, pageNumber},) => {
    try {
      const { data } = await axios.get(`/api/products?keyword=${keyword}&pageNumber=${pageNumber}`);
      return data;
    } catch (error: any) {
      const errorMessage =
        error.response && error.response.data.message
          ? error.response.data.message
          : error.message;
      throw new Error(errorMessage);
    }
  }
);

export const getProductDetails = createAsyncThunk(
  "product/productDetails",
  async (id:string) => {
    try {
      const { data } = await axios.get(`/api/products/${id}`);
      return data;
    } catch (error:any) {
      const errorMessage =
        error.response && error.response.data.message
          ? error.response.data.message
          : error.message;
      throw new Error(errorMessage);
    }
  }
);

export const deleteProduct = createAsyncThunk<IProduct, string, {state: RootState}>(
  "products/deleteProduct",
  async (id, { getState }) => {
    try {
      const { 
        users: { userInfo },
      } = getState();

      if (!userInfo){
        throw new Error('User not signed in')
      }
      const config = {
        headers: {
          "Content-Type": "application/json",
          Authorization: `Bearer ${userInfo.token}`,
        },
      };
      const {data} = await axios.delete(`/api/products/${id}`, config)
      return data
    } catch (error:any) {
      const message =
        error.response && error.response.data.message
          ? error.response.data.message
          : error.message;

      throw new Error(message);
    }
  }
);

interface UpdateInfo {
  _id: string,
  name: string,
  image: string,
  brand: string,
  category: string,
  description: string,
  price: number,
  countInStock: number
}
export const updateProduct = createAsyncThunk<IProduct, UpdateInfo, {state: RootState}>(
  "products/updateProduct",
  async (product, { getState }) => {
    try {
      const {
        users: { userInfo },
      } = getState();

      if (!userInfo){
        throw new Error('User not signed in')
      }
      const config = {
        headers: {
          // "Content-Type": "application/json",
          Authorization: `Bearer ${userInfo.token}`,
        },
      };
      const { data } = await axios.put(
        `/api/products/${product._id}`,
        product,
        config
      );
      return data;
    } catch (error:any) {
      const message =
        error.response && error.response.data.message
          ? error.response.data.message
          : error.message;

      throw new Error(message);
    }
  }
);

export const createProduct = createAsyncThunk<IProduct, void, {state: RootState}>(
  "products/createProduct",
  async (_:void, { getState }) => {
    try {
      const {
        users: { userInfo },
      } = getState();

      if (!userInfo){
        throw new Error("User not logged in ")
      }
      const config = {
        headers: {
          Authorization: `Bearer ${userInfo.token}`,
        },
      };
      const { data } = await axios.post(`/api/products`, {}, config);
      return data;
    } catch (error:any) {
      const errorMessage =
        error.response && error.response.data.message
          ? error.response.data.message
          : error.message;
      throw new Error(errorMessage);
    }
  }
);
interface IReview {
  rating: number,
  comment: string
}
interface createProductReviewParams {
  review: IReview,
  productId: string
}
// <return type, argument type, optional>
export const createProductReview = createAsyncThunk<{}, createProductReviewParams, {state:RootState}>("products/createProductReview", async(reviewParams, {getState}) => {
  const {review, productId} = reviewParams
  try{
    const {
      users: { userInfo },
    } = getState();

  if(!userInfo){
    throw new Error("User not logged in")
  }
    const config = {
      headers: {
        "Content-Type": "application/json",
        Authorization: `Bearer ${userInfo.token}`,
      },
    };
    await axios.post(`/api/products/${productId}/reviews`, review, config);

  }catch(error:any){
    const errorMessage =
    error.response && error.response.data.message
      ? error.response.data.message
      : error.message;
  throw new Error(errorMessage);
  }
})

export const getTopProducts = createAsyncThunk("products/getTopProducts", async() =>{
  try{
    const {data} = await axios.get('/api/products/top')
    return data
  }catch(error:any){
    const errorMessage =
    error.response && error.response.data.message
          ? error.response.data.message
          : error.message;
      throw new Error(errorMessage);
  }
})



interface ProductsState{
  productList: {
    status: string,
    error: any,
    page: number,
    pages: number,
    products: Array<IProduct>
  },
  productDetails: {
    status: string,
    error: any,
    product?: IProduct
  },
  productCreate: {
    status: string,
    error:any,
    product?: IProduct
  },
  productDelete:{
    status:string,
    error: any,
  },
  productUpdate:{
    status: string,
    error: any,
  },
  topProducts:{
    status: string,
    error: any,
    products: Array<IProduct>
  },
  productReview: {
    status: string | undefined,
    error: string | undefined
  }
}


const initialState:ProductsState ={
  productList:{
    status: "",
    error: null,
    page: 1,
    pages: 1,
    products: []
  },
  productDetails:{
    status: "",
    error: null, 
  },
  topProducts:{
    status: "",
    error: null,
    products:[]
  },
  productCreate:{
    status: "",
    error: null,
  }, 
  productDelete:{
    status: "",
    error: null
  },
  productUpdate:{
    status: "",
    error: null
  },
  productReview:{
    status: undefined,
    error: undefined
  }
};
const productsSlice = createSlice({
  name: "products",
  initialState,
  reducers: {
    reset(state){
      return initialState
    }
  },
  extraReducers(builder) {
    builder //list Products
      .addCase(getAllProducts.pending, (state, action) => {
        state.productList.status = "loading";
      })
      .addCase(getAllProducts.fulfilled, (state, action) => {
        state.productList.status = "succeeded";
        const {pages, page, products} = action.payload
        state.productList.pages = pages;
        state.productList.page = page;
        state.productList.products = products
      })
      .addCase(getAllProducts.rejected, (state, action) => {
        state.productList.status = "failed";
        state.productList.error = action.error.message;
      }) 
      .addCase(getProductDetails.pending, (state, action) => {
        state.productDetails.status = "loading";
      })
      .addCase(getProductDetails.fulfilled, (state, action) => {
        state.productDetails.status = "succeeded";
        state.productDetails.product = action.payload
        
      })
      .addCase(getProductDetails.rejected, (state, action) => {
        state.productDetails.status = "failed";
        state.productDetails.error = action.error.message;
      }) 
      .addCase(deleteProduct.pending, (state, action) => {
        state.productDelete.status = "loading";
      })
      .addCase(deleteProduct.fulfilled, (state, action) => {
        state.productDelete.status = "succeeded";
      })
      .addCase(deleteProduct.rejected, (state, action) => {
        state.productDelete.status = "failed";
        state.productDelete.error = action.error.message;
      })
      .addCase(updateProduct.pending, (state, action) => {
        state.productUpdate.status = "loading";    
      })
      .addCase(updateProduct.fulfilled, (state, action) => {
        state.productUpdate.status = "succeeded"
      })
      .addCase(createProduct.pending, (state, action) => {
        state.productCreate.status = "loading";
      })
      .addCase(createProduct.fulfilled, (state, action) => {
        state.productCreate.status = "succeeded";
        state.productCreate.product = action.payload
      })
      .addCase(createProduct.rejected, (state, action) => {
        state.productCreate.status = "failed";
        state.productCreate.error = action.error.message;
      })
      .addCase(getTopProducts.pending, (state, action) => {
        state.topProducts.status = "loading"
      })
      .addCase(getTopProducts.fulfilled, (state, action) => {
        state.topProducts.status = "succeeded"
        state.topProducts.products = action.payload
      })
      .addCase(getTopProducts.rejected,(state, action) => {
        state.topProducts.status = "failed"
        state.topProducts.error = action.error.message
      })
      .addCase(createProductReview.pending, (state, action) => {
        state.productReview.status = "loading"
      })
      .addCase(createProductReview.fulfilled, (state, action) => {
        state.productReview.status = "succeeded"
      })
      .addCase(createProductReview.rejected, (state, action) => {
        state.productReview.status = "failed"
        state.productReview.error = action.error.message
      })
  },
});



export const selectAllProducts = (state: RootState) => state.products.productList
export const selectProductDetails = (state: RootState) => state.products.productDetails
export const {reset} = productsSlice.actions

export default productsSlice.reducer;
