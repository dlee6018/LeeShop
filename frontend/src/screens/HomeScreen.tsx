import React, { useEffect } from "react";
import { Link } from "react-router-dom";
import { useDispatch, useSelector } from "react-redux";
import { Row, Col } from "react-bootstrap";
import Product from "../components/Product";
import Message from "../components/Message";
import Loader from "../components/Loader";
import Paginate from "../components/Paginate";
import ProductCarousel from "../components/ProductCarousel";
import Meta from "../components/Meta";
import {
  getProductsStatus,
  getProductsError,
  getAllProducts,
} from "../features/products/productSlice";
import { RootState } from "../store";


type HomeScreenProps = {
  match:any
}
const HomeScreen = ({ match }:HomeScreenProps) => {
  const keyword = match.params.keyword || "";

  const pageNumber = match.params.pageNumber || 1;

  const dispatch = useDispatch();

  const products = useSelector((state:RootState) => state.products.productList);
  const status = useSelector(getProductsStatus);
  const error = useSelector(getProductsError);
  const page = useSelector((state:RootState) => state.products.page);
  const pages = useSelector((state:RootState) => state.products.pages);

  const searchInfo = {
    keyword,
    pageNumber,
  };
  useEffect(() => {
    dispatch(getAllProducts({ keyword, pageNumber }));
  }, [dispatch, keyword, pageNumber]);
  return (
    <>
      <Meta />
      {!keyword.length ? (
        <ProductCarousel />
      ) : (
        <Link to="/" className="btn btn-light">
          Go Back
        </Link>
      )}
      <h1>Latest Products</h1>
      {status == "loading" ? (
        <Loader />
      ) : status == "error" ? (
        <Message variant="danger">{error}</Message>
      ) : (
        <>
          <Row>
            {products.map((product) => (
              <Col key={product._id} sm={12} md={6} lg={4} xl={3}>
                <Product product={product} />
              </Col>
            ))}
          </Row>
          <Paginate
            pages={pages}
            page={page}
            keyword={keyword ? keyword : ""}
          />
        </>
      )}
    </>
  );
};

export default HomeScreen;
