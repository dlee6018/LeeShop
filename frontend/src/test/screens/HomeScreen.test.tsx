import React from 'react'
import { rest } from 'msw'
import { setupServer } from 'msw/node'
import { fireEvent, screen } from '@testing-library/react'
// We're using our own custom render function and not RTL's render.
import { renderWithProviders } from '../../types/test-utils'
import HomeScreen from '../../screens/HomeScreen'
import product from '../../../../backend/data/products'
import store from '../../store'
import { getAllProducts } from '../../features/products/productSlice'
// We use msw to intercept the network request during the test,
// and return the response 'John Smith' after 150ms
// when receiving a get request to the `/api/user` endpoint

export const handlers = [
  rest.get('/api/', (req, res, ctx) => {
    return res(ctx.json(product), ctx.delay(150))
  })
]

const server = setupServer(...handlers)

// Enable API mocking before tests.
beforeAll(() => server.listen())

// Reset any runtime request handlers we may add during the tests.
afterEach(() => server.resetHandlers())

// Disable API mocking after the tests are done.
afterAll(() => server.close())

const matchProps = {
  params:{
    keyword: ""
  }
}

test('fetches & receives a user after clicking the fetch user button', async () => {
  renderWithProviders(<HomeScreen match = {matchProps}/>, {
    preloadedState:{
      products: product as any
    }
  })


  store.dispatch(getAllProducts({keyword:"", pageNumber:1}))
  // expect(screen.getByText("Latest Products")).toBeInTheDocument()
  expect(<HomeScreen match = {matchProps}/>).toMatchSnapshot()
})