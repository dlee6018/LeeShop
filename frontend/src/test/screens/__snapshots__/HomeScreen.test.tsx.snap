// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`fetches & receives a user after clicking the fetch user button 1`] = `
<HomeScreen
  match={
    Object {
      "params": Object {
        "keyword": "",
      },
    }
  }
/>
`;
